### PR TITLE
Add org-transclusion-toggle

### PR DESCRIPTION
**Pull request by joshcho**
Added a prefix-insert function and a toggle function. Add org-transclusion-boundary which when set to t adds a blank line to end of all transcluded text.

### CODE REVIEW

File: org-transclusion.el [hunk 1]
```diff
@@ -254,6 +254,12 @@ specific keybindings; namely:
 
 (defvar org-transclusion-yank-remember-user-excluded-props '())
 
+(defvar org-transclusion-latex-preview-on-toggle nil
+  "Call org-latex-preview after transclusion-toggle.")
+
+(defvar org-transclusion-boundary nil
```
Comment (nobiot) [inline org-transclusion.el:260]: For this (adding an extra empty line at the end of transclusion, I'd rather revert this change I made:

https://github.com/nobiot/org-transclusion/blob/3fb2a35e79c8542e2b9b7d03bbd75a9ed278e61a/org-transclusion.el#L1160-L1167

File: org-transclusion.el [hunk 2]
```diff
@@ -562,6 +568,35 @@ the rest of the buffer unchanged."
       (move-marker marker nil) ; point nowhere for GC
       list)))
 
+(defun org-transclusion-toggle ()
```
Comment (nobiot) [inline org-transclusion.el:571]: What does this do that add/remove cannot do?

It looks to be that you would like to call `org-latex-preview`. Perhaps we could consider adding a hook within `org-transclusion-add` after the content has been inserted?

File: org-transclusion.el [hunk 3]
```diff
@@ -562,6 +568,35 @@ the rest of the buffer unchanged."
...
+  (interactive)
+  (condition-case nil
+      (progn
+        (org-transclusion-remove)
+        (org-transclusion-remove-all))
+    (error (org-transclusion-add-all)))
+  (if org-transclusion-latex-preview-on-toggle
+      (org-latex-preview '(16))))
+
+(defun org-transclusion-prefix-insert ()
```
Comment (nobiot) [inline org-transclusion.el:594]: `prefix-insert` seems to have the same intention as `org-transclusion-make-from-link` does -- create a `#+transclude` link out of a normal link. 

What would it do if the link is not in the beginning of a paragraph? For instance, it's common to have a paragraph like this:

> His work has had a considerable influence on mathematics, logic, set theory, [[file:linguistics.org][linguistics]], artificial intelligence, cognitive science, computer science, and various areas of analytic philosophy, especially philosophy of mathematics, philosophy of language, epistemology and metaphysics

**Comment by nobiot** [comment]
Thank you for putting together a PR. It has more than one elements to discuss, so I have left comments with using the review feature of GitHub.

Before discussing the details of the PR, there is one big thing we need to clarify: FSF copyright paperwork.

Org-transclusion is published on ELPA, which means that any contributor with more than 15 lines of code will need to have signed the FSF copyright paperwork.
It's mentioned in [Contributing](https://github.com/nobiot/org-transclusion#contributing) section of README

> Org-transclusion is part of GNU ELPA and thus copyrighted by the Free Software Foundation (FSF). This means that anyone who is making a substantive code contribution will need to “assign the copyright for your contributions to the FSF so that they can be included in GNU Emacs” (Org Mode website).

Have you done this before? If not, are you willing to and able to go through this process?

File: org-transclusion.el (hunk 1 above)
Comment (joshcho) [reply org-transclusion.el:260]: That makes sense, this looks good to me.

File: org-transclusion.el (hunk 2 above)
Comment (joshcho) [reply org-transclusion.el:571]: I don't think toggle adds anything new other than add/remove, but it's the main way in which I ended up using transclusion: as a way to preview content quickly. I think something like this should at least be natively there, even if it doesn't necessarily "add" a functionality.

The hook makes sense, as latex previews get broken a lot when I toggle links often.

File: org-transclusion.el (hunk 3 above)
Comment (joshcho) [reply org-transclusion.el:594]: I wasn't aware of that function, thanks for referencing it.

**Comment by joshcho** [comment]
I read about FSF copyright paperwork in another pull request, so I have already gone through the process. I have not received a reply from FSF yet, but I sent an email.

File: org-transclusion.el (hunk 3 above)
Comment (joshcho) [reply org-transclusion.el:594]: Actually, this may be specific to my own purposes, but as you may have seen in my #120 screenshot, I have links by themselves on a line. I essentially want to be able to preview and unpreview things, without necessarily having `#+transclude:` everywhere. `prefix-insert` essentially is part of `org-transclusion-toggle` in allowing links to stay as they are (without `#+transclude:`), but allow preview and unpreview of links.

The analogy is I have Luhmann's slipbox, with each link representing the "title" of the notecards viewable from above, but you can "scroll through" the slipbox by previewing each entry.

Compared to this, `org-transclusion-make-from-link` actually just makes a new link, which seems to be for a different use case.

Comment (joshcho) [reply org-transclusion.el:594]: With regard to the example you referenced, I had thought about that and how that could play with `org-transclusion-toggle` in general. I think, again for my own purposes, it would be nice to have something like `org-transclusion-make-from-link` below the paragraph when `org-transclusion-toggle` is called, as `org-transclusion-toggle` for lines with text included doesn't work at the moment.

**Comment by joshcho** [comment]
Another thought I had: I plan on adding more things that are more specific to a specific workflow (emulating Luhmann slipbox). It may make more sense to move all this UX-like functionality out of `org-transclusion`, and put it in a new package or sth like `org-slipbox` that uses `org-transclusion` as a package.

Also, just wanted to say thank you for this package! It's something that I would have needed to write at some point (for my own uses and satisfaction), but it's already written in excellent condition.

File: org-transclusion.el (hunk 2 above)
Comment (nobiot) [reply org-transclusion.el:571]: I've implemented the hook and created a new feature branch for the hook. 
Have a look at PR #124.

I don't use `latex` etc. and can't get the latex preview working, so I'd appreciate it if you could test it on your end.

File: org-transclusion.el [hunk 4]
```diff
@@ -942,12 +977,15 @@ based on the following arguments:
          (end-mkr)
          (ov-src (text-clone-make-overlay sbeg send sbuf)) ;; source-buffer overlay
          (tc-pair ov-src)
-         (content content))
+         (content content)
+         (at-end-of-buffer (eq (point-at-eol) (point-max))))
```
Comment (nobiot) [inline org-transclusion.el:981]: @joshcho with reverting the commit for "removing the extra line", I don't think any of these boundary stuff is necessary. Agree?

File: org-transclusion.el (hunk 3 above)
Comment (nobiot) [reply org-transclusion.el:594]: > again for my own purposes, it would be nice to have something like org-transclusion-make-from-link below the paragraph when org-transclusion-toggle is called, as org-transclusion-toggle for lines with text included doesn't work at the moment.

I think you would need to decide whether or not you would like the preview feature to be for "your own purpose" or something to be shared by others.

**Comment by nobiot** [comment]
> Also, just wanted to say thank you for this package! It's something that I would have needed to write at some point (for my own uses and satisfaction), but it's already written in excellent condition.

Thank you for your kind words. Much appreciated.

> Another thought I had: I plan on adding more things that are more specific to a specific workflow (emulating Luhmann slipbox). It may make more sense to move all this UX-like functionality out of `org-transclusion`, and put it in a new package or sth like `org-slipbox` that uses `org-transclusion` as the core engine.

Creating another package for others to use is one idea. 
Another possibility is to share `org-slipbox` as an extension to `org-transclusion`, if you think it is not big enough to justify a separate package. 

But before talking about this idea further, I would like you to do the following, if you can:

1. Test the new `after-add-hook` for `org-latex-preview`
2. Decide if you would like to keep the toggle / preview feature as your personal one, or pursue this PR further to be shared with others (maybe as an extension, like `org-transclusion-indent-mode` and others)
3. If you would like to pursue, remove the code for `boundary` and `org-latex-preview` -- I believe that's no longer needed

File: org-transclusion.el (hunk 4 above)
Comment (joshcho) [reply org-transclusion.el:981]: Yup, that sounds good.

Comment (joshcho) [reply org-transclusion.el:981]: So for my own purposes, should I just revert that change within `org-transclusion.el`? Or is there a more elegant way of customization?

**Comment by joshcho** [comment]
In response to the three points:

1. I have tested it and have shared the results in the corresponding pull request.
2. I think I want to keep tinkering with it; there are more moving parts than I had imagined and aesthetics I would like to maintain.
3. In that case, I believe the only change necessary is the post hook, which you have put in another pull request. I will be closing accordingly.

File: org-transclusion.el (hunk 1 above)
Comment (joshcho) [reply org-transclusion.el:260]: Should this be active in the feature branch? I do not see any changes regarding this point.

Comment (nobiot) [reply org-transclusion.el:260]: It’s already in the main and merged with the feature branch too
